### PR TITLE
HW02: solutions by Tatsiana Zenkovich

### PR DESCRIPTION
**Pull request by TanjaZzz**

### CODE REVIEW

File: HW02/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_02_t_01.rb [hunk 1]
```diff
@@ -0,0 +1,10 @@
+# frozen_string_literal: true
+
+lines = []
+File.open('log.txt') do |review_file|
+  lines = review_file.readlines
+end
+relevant_line = []
+lines.each do |line|
+  relevant_line << line if line.include?('error')
+end
```
Comment (B0RlS) [inline HW02/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_02_t_01.rb]: Your code output is incorrect, will return:
_" ["10.6.246.103 - - [23/Apr/2018:20:30:39 +0300] \"POST /grid/2/messages HTTP/1.1\" 200 48 0.0498\n", "10.6.246.101 - - [23/Apr/2018:20:30:42 +0300] \"POST /grid/2/event HTTP/1.1\" 200 - 0.2277\n", "2018-04-23 20:30:42: SSL error, peer: 10.6.246.101, peer cert: , #<Puma::MiniSSL::SSLError: System error: Undefined error: 0 - 0>\n", "10.6.246.101 - - [23/Apr/2018:20:29:39 +0300] \"POST /grid/2/messages HTTP/1.1\" 200 48 0.0290"]"_

Could you please create 2 methods:
- read_file method (return the contents of the string)
- find_error method accepting string in arguments and return ' ' or 'first line with an error'

File: HW02/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_02_t_01.rb [hunk 2]
```diff
@@ -0,0 +1,11 @@
+# frozen_string_literal: true
+
+lines = []
+File.open('log.txt') do |review_file|
+  lines = review_file.readlines
+end
```
Comment (vladhilko) [inline HW02/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_02_t_01.rb]: You can use one line syntax 
```suggestion
lines = File.readlines('log.txt')
```

File: HW02/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_02_t_01.rb [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+relevant_line = []
+lines.each do |line|
+  relevant_line << line if line.include?('error')
+end
```
Comment (vladhilko) [inline HW02/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_02_t_01.rb]: You can use `.select` here https://apidock.com/ruby/Array/select
```suggestion
lines.select { ... } 
```

File: HW02/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_02_t_01.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (TanjaZzz) [reply HW02/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_02_t_01.rb]: Thank you, I fixed that

File: HW02/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_02_t_01.rb (hunk 3 above)
Comment (TanjaZzz) [reply HW02/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_02_t_01.rb]: I corrected both, as you suggested

Comment (TanjaZzz) [reply HW02/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_02_t_01.rb]: Thank you!

File: HW02/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_02_t_01.rb [hunk 4]
```diff
@@ -0,0 +1,6 @@
+def task1
+  lines = File.readlines('log.txt')
+  puts lines.select { |line| line.include?('error') }.first
```
Comment (vladhilko) [inline HW02/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_02_t_01.rb]: Normally, we don't use `puts` inside methods, a method just should return a value

File: HW02/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_02_t_01.rb [hunk 5]
```diff
@@ -0,0 +1,6 @@
...
+end
+
+task1
```
Comment (vladhilko) [inline HW02/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_02_t_01.rb]:
```suggestion
puts task1
```

File: HW02/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_02_t_02.rb [hunk 6]
```diff
@@ -0,0 +1,18 @@
+# frozen_string_literal: true
+
+lines = []
+
+File.open('log.txt') do |review_file|
+  lines = review_file.readlines
+end
+
+re = []
```
Comment (vladhilko) [inline HW02/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_02_t_02.rb]: incommunicative name

File: HW02/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_02_t_02.rb [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+lines.each do |i|
+  ip = i[/\d+\.\d+\.\d+\.\d+/]
+  date = i[%r{\[\d+/\w+/\d{4}\:\d{2}\:\d{2}\:\d{2}\s\+\d+\]}]
+  adr = i[/T .*H/]
```
Comment (vladhilko) [inline HW02/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_02_t_02.rb]: uncommunicative name, please, use the whole name
```suggestion
  address = i[/T .*H/]
```

File: HW02/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_02_t_02.rb [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+lines = []
+
+File.open('log.txt') do |review_file|
+  lines = review_file.readlines
+end
+
+re = []
+lines.each do |i|
+  ip = i[/\d+\.\d+\.\d+\.\d+/]
+  date = i[%r{\[\d+/\w+/\d{4}\:\d{2}\:\d{2}\:\d{2}\s\+\d+\]}]
```
Comment (vladhilko) [inline HW02/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_02_t_02.rb]: We can move these variables to CONSTANT
```suggestion
  IP = i[/\d+\.\d+\.\d+\.\d+/]
  DATE = i[%r{\[\d+/\w+/\d{4}\:\d{2}\:\d{2}\:\d{2}\s\+\d+\]}]
```

File: HW02/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_02_t_01.rb (hunk 4 above)
Comment (vladhilko) [reply HW02/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_02_t_01.rb]: and we can use `.find` to return only one record

Comment (TanjaZzz) [reply HW02/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_02_t_01.rb]: why .first is not okay ?

Comment (vladhilko) [reply HW02/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_02_t_01.rb]: `first` is okay, but `.find` is the shortest way to do the same

Comment (vladhilko) [reply HW02/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_02_t_01.rb]: https://apidock.com/ruby/Enumerable/find

Comment (TanjaZzz) [reply HW02/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_02_t_01.rb]: got it, thanks!) it works faster, as don't check all the array, as select do.

File: HW02/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_02_t_02.rb (hunk 8 above)
Comment (TanjaZzz) [reply HW02/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_02_t_02.rb]: Why do you suggest moving them to constant? what's wrong with leaving it as a variable?

Comment (vladhilko) [reply HW02/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_02_t_02.rb]: It's just best practice. If we aren't going to change variables, we should move them to constants

https://stackoverflow.com/questions/29927103/when-to-use-constants-instead-of-instance-variables-in-ruby

File: HW02/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_02_t_01.rb (hunk 5 above)
Comment (TanjaZzz) [reply HW02/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_02_t_01.rb]: improved!

File: HW02/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_02_t_02.rb (hunk 8 above)
Comment (TanjaZzz) [reply HW02/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_02_t_02.rb]: I changed as suggested, and now get many warnings during a code execution like this. Should I just ignore it?
"tatsiana_zenkovich_hw_02_t_02.rb:9: warning: already initialized constant IP
tatsiana_zenkovich_hw_02_t_02.rb:9: warning: previous definition of IP was here
tatsiana_zenkovich_hw_02_t_02.rb:10: warning: already initialized constant DATE
tatsiana_zenkovich_hw_02_t_02.rb:10: warning: previous definition of DATE was here
tatsiana_zenkovich_hw_02_t_02.rb:11: warning: already initialized constant ADDRESS"

Comment (TanjaZzz) [reply HW02/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_02_t_02.rb]: despite those warnings, the code makes correct output

Comment (vladhilko) [reply HW02/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_02_t_02.rb]: You need to move these constants outside of `each` block, there's no sense to redefine them everytime

File: HW02/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_02_t_02.rb [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+
+File.open('log.txt') do |review_file|
+  lines = review_file.readlines
+end
+
+found = []
+lines.each do |i|
+  IP = i[/\d+\.\d+\.\d+\.\d+/]
+  DATE = i[%r{\[\d+/\w+/\d{4}\:\d{2}\:\d{2}\:\d{2}\s\+\d+\]}]
+  ADDRESS = i[/T .*H/]
```
Comment (vladhilko) [inline HW02/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_02_t_02.rb]:
```suggestion
IP = i[/\d+\.\d+\.\d+\.\d+/]
DATE = i[%r{\[\d+/\w+/\d{4}\:\d{2}\:\d{2}\:\d{2}\s\+\d+\]}]
ADDRESS = i[/T .*H/]
found = []
lines.each do |i|

```

Comment (TanjaZzz) [reply HW02/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_02_t_02.rb]: Hm. Then should I also remove "i" in each constant, as we didn't define it before, right? and then there's a problem with "no implicit conversion of String into Array (TypeError)"
May be I still leave them as variables?

Comment (vladhilko) [reply HW02/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_02_t_02.rb]: ahhh, I see 😅 then we need to do something like this
```ruby
IP = /\d+\.\d+\.\d+\.\d+/
lines.each do |i|
  ip = i[IP]
```

Comment (vladhilko) [reply HW02/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_02_t_02.rb]: sorry, my bad 😅

Comment (TanjaZzz) [reply HW02/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_02_t_02.rb]: looks like I also have to .freeze CONSTANTS then :)
I corrected all as you suggested.

File: HW02/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_02_t_04.rb [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+puts 'Please, type your string'
+input = gets.chomp
+
+numbers_count = 0
+string_count = 0
+
+input.split('').each do |x|
+  numbers_count += 1 if x =~ /\d/
+  string_count += 1 if x =~ /[a-z]|[A-Z]/
+end
```
Comment (vladhilko) [inline HW02/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_02_t_04.rb]: interesting 👍

File: HW02/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_02_t_03.rb [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+# messages
+
+def task3
+  lines = read_file('log.txt')
+  relevant_line = parsing(lines)
+  check_the_line(relevant_line)
+  time_between = count_time(relevant_line)
+  names = names_array(relevant_line)
+  time_between = check_single_calling_core(time_between)
+  final_array = creating_output(time_between, names)
```
Comment (vladhilko) [inline HW02/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_02_t_03.rb]: This is the right approach 💯 You code should tell a story

File: HW02/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_02_t_03.rb [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+  puts final_array
+end
+
+def read_file(file_name)
+  lines = []
+  File.open(file_name) do |review_file|
+    lines = review_file.readlines
+  end
+  lines
+end
```
Comment (vladhilko) [inline HW02/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_02_t_03.rb]:
```suggestion
def read_file(file_name)
  File.readlines(file_name)
end
```

File: HW02/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_02_t_03.rb [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+
+def parsing(lines)
+  relevant_line = []
+  lines.each do |line|
+    relevant_line << line if line.include?('Calling core with action')
+  end
+  relevant_line
+end
```
Comment (vladhilko) [inline HW02/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_02_t_03.rb]:
```suggestion
def parsing(lines)
  lines.select { |line| line.include?('Calling core with action') }
end
```

File: HW02/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_02_t_03.rb [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+
+# check if the log contains any Calling core with action
+def check_the_line(relevant_line)
+  puts '0' && exit if relevant_line == []
+end
+
+# count time for each Calling core with action
+
+def count_time(relevant_line)
+  require 'time'
```
Comment (vladhilko) [inline HW02/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_02_t_03.rb]: Please, don't do it inside methods, could you move it to the beginning of the file?

File: HW02/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_02_t_03.rb [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+end
+
+def names_array(relevant_line)
+  names = []
+  relevant_line.each do |new_line|
+    name = new_line.split(' ').last
+    names << name
+  end
+  names
+end
```
Comment (vladhilko) [inline HW02/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_02_t_03.rb]:
```suggestion
def names_array(relevant_line)
  relevant_line.map { |new_line| new_line.split(' ').last }
end
```

File: HW02/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_02_t_03.rb (hunk 14 above)
Comment (TanjaZzz) [reply HW02/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_02_t_03.rb]: Sure. I corrected all, as you suggested.

File: HW02/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_02_t_04.rb (hunk 10 above)
Comment (TanjaZzz) [reply HW02/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_02_t_04.rb]: :)

**Pull request merged**